app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt

```diff
@@ -0,0 +1,178 @@
+    }
+
+    /**
+     * Removes the keys in [deleted] from cache
+     * Puts the [updatedOrAdded] items into cache only if the containing tile is already cached
+     */
+    fun update(updatedOrAdded: Iterable<T> = emptyList(), deleted: Iterable<K> = emptyList()) = synchronized(this) {
+        val deleteKeys = deleted.asSequence() + updatedOrAdded.map { it.getKey() }
+        for (key in deleteKeys) {
+            val item = byKey.remove(key) ?: continue
+            byTile[item.getTilePos()]?.remove(item)
+        }
+        // items must be removed before they are re-added because e.g. position may have changed
+        for (item in updatedOrAdded) {
+            val oldItem = byKey.remove(item.getKey()) ?: continue
+            byTile[oldItem.getTilePos()]?.remove(oldItem)
+        }
+        // items are only added if the tile they would be sorted into is already cached, because all
+        // tiles that are in cache must be complete
+        for (item in updatedOrAdded) {
+            val tile = byTile[item.getTilePos()]
+            if (tile != null) {
+                tile.add(item)
+                byKey[item.getKey()] = item
+            }
+        }
+    }
+
+    /**
+     * Replaces all tiles fully contained in the [bbox] with the given [items].
+     * Any tile only partially contained in the [bbox] is removed (because all tiles cached must be
+     * complete).
+     * If the number of tiles exceeds maxTiles, only the first maxSize tiles are cached.
+     */
+    fun replaceAllInBBox(items: Collection<T>, bbox: BoundingBox) = synchronized(this) {
+        val tiles = bbox.asListOfEnclosingTilePos()
+        val (completelyContainedTiles, incompleteTiles) = tiles.partition { it.asBoundingBox(tileZoom).isCompletelyInside(bbox) }
+        if (incompleteTiles.isNotEmpty()) {
+            Log.w(TAG, "bbox does not align with tiles, clearing incomplete tiles from cache")
+            for (tile in incompleteTiles) {
+                removeTile(tile)
+            }
+        }
+        replaceAllInTiles(items, completelyContainedTiles)
+
+        trim()
+    }
+
+    // may add tiles, but does not call trim()
+    private fun replaceAllInTiles(items: Collection<T>, tiles: Collection<TilePos>) {
+        // create / replace tiles
+        for (tile in tiles) {
+            removeTile(tile)
+            byTile[tile] = HashSet()
+        }
+/*        for (item in items) {
+            val tile = byTile[item.getTilePos()] ?: continue
+            tile.add(item)
+            byKey[item.getKey()] = item
+        }*/
+        val bboxByTile = byTile.keys.associateWith { it.asBoundingBox(tileZoom) }
+        items@for (item in items) {
+            val pos = item.getPosition()
+            for (tile in tiles) {
+                if (pos in bboxByTile[tile]!!) {
+                    byTile[tile]!!.add(item)
+                    byKey[item.getKey()] = item
+                    continue@items
+                }
+            }
+        }
```

Comment:
Here is some commented out code left, not sure what this is about.

Reply:
The commented code is the version removed as part of https://github.com/streetcomplete/StreetComplete/pull/4125#discussion_r970713390.
If the (potential) bug is fixed, it could replace the nested loop below.

But there is no need to keep the commented code...

Reply:
removed